### ISSUE TITLE
Does it make sense to create empty optional instances?

### ISSUE THREAD

**Issue by bali182** [JasonStorey/Optional.js#7]
First, cheers for porting this to JS! 

This ticket is an idea on how you could make the library more performant using value classes and a single value for absent. If you don't like this idea (because in javascript everything is mutable and users could mutate the absent whenever they want) I completely understand, just wanted to bring it to your attention. If you are interested, guava and scala optionals are implemented this way as well.

For the sake of simplicity I'm using ES6 classes, but these could easily transformed to ES5 like the current codebase.

First we have an abstract `Optional` class, the methods don't do anything, just throw if directly instantiated (or methods could be omitted completely):

```js
export class Optional {
  get() {
    throw new Error('not implemented')
  }
  
  filter() {
    throw new Error('not implemented')
  }
  
  map() {
    throw new Error('not implemented')
  }
}
```

Now we can have an empty or absent (or whatever) optional, which NEVER checks with any branches if there is a value, since we created this class for the purpose of NOT holding a value.

```js
class _Absent extends Optional {
  get() {
    throw new Error('get() called on Absent')
  }
  
  filter() {
    return this
  }
  
  map() {
    return this
  }
}
```

Notice that I didn't export the class, since we don't want anyone to instantiate this. We only need a single instance, which we can export:

```js
export const Absent = new _Absent()
```
Now we can do the same for a subclass which ALWAYS holds a value:

```js
export class Present extends Optional {
  constructor(value) {
    super()
    this.value = value
  }
  
  get() {
    return this.value
  }
  
  filter(fn) {
    return fn(this.value) ? this : Absent
  }
  
  map(fn) {
    return new Present(fn(this.value))
  }
}
```

Now we can add a static factory method to optional, which does the logic, so the classes themselves don't have to:

```js
export class Optional {
  // ... rest of the methods
  static ofNullable(value) {
    return isNull(value) ? Absent : new Optional(value)
  }
}
```

You could basically keep the API the same, except under the hood there is no null checking EVER except when calling `Optional.ofNullable`. Let me know what you think!

**Comment by JasonStorey** [comment]
Nice idea. I'm still thinking it through.

What about the 'map' and 'flatMap' instance methods? They'd still have to do the null check, right?

**Comment by whirlwin** [comment]
I think `flatMap` should have the null check, but why `map`?

```js
Optional.ofNullable('hello')
    .map(value => null)
    .get(); // Should in fact be *null* in my opinion
```

**Comment by bali182** [comment]
Yes, I wouldn't check in `map`, if it's `null` or not, thats the users responsibility. In `flatMap` I woud do a check, but throw if the returned value is `null`

**Comment by stavalfi** [comment]
@whirlwin `get` should throw an error if he sees null. 
@bali182 `map` must throw an error in case of a null because this library is used to remove all undefined and nulls from your project. there is no reason to use Optional.js and still use nulls. Even worse, represent null in Optional type. **What does it even mean? Any usecase?**


In any way, I think the overall idea is nice and embrases moands (functional programming).

Maybe.js is implemented that way. The only difference between maybe.js and optional.js is that maybe.js can't represent functions because it invokes them (which is the reason I moved from maybe.js to optional.js).
